lib/notifications/notification_server.ex

```diff
@@ -0,0 +1,116 @@
+defmodule Notifications.NotificationServer do
+  use GenServer
+
+  alias Notifications.Notification
+  alias Realtime.BlockWaiver
+  alias Schedule.Block
+
+  require Logger
+
+  @type __MODULE__ :: {schedule_data_mod :: module()}
+
+  defstruct [:schedule_data_mod]
+
+  # Client
+
+  @spec default_name() :: GenServer.name()
+  def default_name(), do: Notifications.NotificationServer
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []) do
+    name = Keyword.get(opts, :name, __MODULE__)
+    schedule_data_mod = Keyword.get(opts, :schedule_data_mod, Schedule)
+    GenServer.start_link(__MODULE__, %{schedule_data_mod: schedule_data_mod}, name: name)
+  end
+
+  @spec new_block_waivers(BlockWaiver.block_waivers_by_block_key(), GenServer.server()) :: :ok
+  def new_block_waivers(new_waivers_by_block_key, server \\ default_name()) do
+    GenServer.cast(server, {:new_block_waivers, new_waivers_by_block_key})
+  end
+
+  # Server
+
+  @impl GenServer
+  def init(initial_state) do
+    schedule_data_mod = Map.fetch!(initial_state, :schedule_data_mod)
+    {:ok, %__MODULE__{schedule_data_mod: schedule_data_mod}}
+  end
+
+  @impl true
+  def handle_cast({:new_block_waivers, new_block_waivers_by_block_key}, state) do
+    new_notifications =
+      new_block_waivers_by_block_key
+      |> convert_new_block_waivers_to_notifications(state.schedule_data_mod)
+
+    if !Enum.empty?(new_notifications) do
+      Logger.warn(
+        "NotificationServer created new notifications new_notifications=#{
+          inspect(new_notifications)
+        }"
+      )
+    end
+
+    {:noreply, state}
+  end
+
+  @spec convert_new_block_waivers_to_notifications([BlockWaiver.t()], module()) :: [
+          Notification.t()
+        ]
+  defp convert_new_block_waivers_to_notifications(new_block_waivers, schedule_data_mod) do
+    new_block_waivers
+    |> Enum.flat_map(fn {block_key, block_waivers} ->
+      Enum.map(
+        block_waivers,
+        &convert_block_waiver_to_notification(block_key, &1, schedule_data_mod)
+      )
+    end)
+    |> Enum.filter(& &1)
```

Comment:
Not sure what you mean by "outside the flatmap, everything should already be lists." The `filter` works where it is, it would also work inside the `flat_map`--six of one, half a dozen of the other here.

Reply:
Oh yeah, for some reason I was thinking that the nil values would break the flattening, but they don't. Nevermind.